z3.ads

```diff
@@ -434,15 +434,22 @@ package Z3 is  --  GCOV_EXCL_LINE
 
 private
 
-   type Config is
-      record
-         Data : z3_api_h.Z3_config;
-      end record;
+   type Reference_Counter is access Natural;
 
-   type Context is
-      record
-         Data : z3_api_h.Z3_context;
-      end record;
+   type Context is new Ada.Finalization.Controlled with record
+      Data      : z3_api_h.Z3_context;
+      Config    : z3_api_h.Z3_config;
+      Ref_Count : Reference_Counter;
```

Comment:
What is the reason for `Ref_Count` being an access type?

Reply:
When `Context` is copied, `Adjust` will be called on the copy. Due to the access type all copies hold a reference on the same reference counter. If it was just an integer all copies would contain their own reference counter and there would be no way to keep them synchronized.

Reply:
I see.